src/AggregateFunctions/AggregateFunctionDeepMergeJSON.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include <map>
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <Columns/ColumnObject.h>
+#include <DataTypes/DataTypeObject.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <base/StringRef.h>
+#include <Common/Arena.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+/// Reasonable limits to prevent memory abuse
+constexpr size_t MAX_JSON_MERGE_PATHS = 10000;
+constexpr size_t MAX_JSON_MERGE_PATH_LENGTH = 1000;
+constexpr size_t MAX_JSON_MERGE_TOTAL_SIZE = 100_MiB;
+
+/// JSON unset marker key
+constexpr const char * UNSET_KEY = "$unset";
```

Comment:
I suggest making this key as an optional parameter of an aggregate function, so if it's not set we don't need to handle deletes at all